### PR TITLE
Jar-In-Jar

### PR DESCRIPTION
**Pull request by marchermans**
### Core features
- Loading of included jars from mod jars
- Update to SPI 5.x
- Introduce early mod deduplication
- Introduce dependency locators base implementation
- Allow providers to throw early loading exceptions and handle them gracefully

### New libraries:
- JarJar (Jar in Jar dependency resolution, metadata specification and filesystems (https://github.com/MinecraftForge/JarJar))

### New subsystems:
- Jar in Jar loading

### New test cases:
- No new test mods, but the test source set now contains a recompiled simple mod that gets loaded at the start to validate that it can load.

#### Metadata format:
```json
{
  "jars": [
    {
      "identifier": {
        "group": "net.minecraftforge.jij",
        "artifact": "JijTestMod"
      },
      "version": {
        "range": "[1.0,)",
        "artifactVersion": "1.2"
      },
      "path": "META-INF/jarjar/ForgeJiJTestMod-1.2.jar"
    }
  ]
}
```
The metadata file needs to be located in the `META-INF/jarjar` directory as per JarJar specifications.

#### Loaded mods:
image

#### Error screen when dependency resolution fails:
image

### CODE REVIEW

**Comment by sciwhiz12** [comment]
As a note, this PR's equivalent for the Latest version (1.19.x), #8715, has been merged. That makes this PR into an LTS backport PR.

**Comment by marchermans** [comment]
Yeah this also needs some cleanup still and some fixes

**Comment by marchermans** [comment]
Yeah this also needs some cleanup still and some fixes

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractJarFileModProvider.java [hunk 1]
```diff
@@ -6,30 +6,19 @@
...
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
 import java.util.function.Consumer;
 import java.util.function.Function;
-import java.util.jar.Manifest;
-import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
-public abstract class AbstractJarFileLocator extends AbstractModLocator {
```
Comment (covers1624) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractJarFileModProvider.java]: Binary breaking class name change.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java [hunk 2]
```diff
@@ -24,8 +24,9 @@
 import java.util.function.Function;
 import java.util.jar.Manifest;
 
-public abstract class AbstractModLocator implements IModLocator {
```
Comment (covers1624) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java]: Binary breaking class name change.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ClasspathLocator.java [hunk 3]
```diff
@@ -24,7 +24,8 @@
 import java.util.Map;
 import java.util.stream.Stream;
 
-public class ClasspathLocator extends AbstractJarFileLocator {
```
Comment (covers1624) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ClasspathLocator.java:27]: Unnecessary formatting change.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/MavenDirectoryLocator.java [hunk 4]
```diff
@@ -15,7 +15,8 @@
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
-public class MavenDirectoryLocator extends AbstractJarFileLocator {
```
Comment (covers1624) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/MavenDirectoryLocator.java:18]: Unnecessary formatting change.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/MinecraftLocator.java [hunk 5]
```diff
@@ -23,7 +23,8 @@
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
-public class MinecraftLocator extends AbstractModLocator {
```
Comment (covers1624) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/MinecraftLocator.java:26]: Unnecessary formatting change.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModInfo.java [hunk 6]
```diff
@@ -185,7 +185,6 @@ public ModVersion(final IModInfo owner, final IConfigurable config) {
                     .orElse(DependencySide.BOTH);
         }
 
-
```
Comment (covers1624) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModInfo.java]: Unnecessary formatting change.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModsFolderLocator.java [hunk 7]
```diff
@@ -23,7 +23,8 @@
 /**
  * Support loading mods located in JAR files in the mods folder
  */
-public class ModsFolderLocator extends AbstractJarFileLocator {
```
Comment (covers1624) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModsFolderLocator.java:26]: Unnecessary formatting change.

File: src/main/java/net/minecraftforge/client/gui/ModListScreen.java [hunk 8]
```diff
@@ -451,7 +453,7 @@ public void upload() {
 
         if (vercheck.status() == VersionChecker.Status.OUTDATED || vercheck.status() == VersionChecker.Status.BETA_OUTDATED)
             lines.add(ForgeI18n.parseMessage("fml.menu.mods.info.updateavailable", vercheck.url() == null ? "" : vercheck.url()));
-        lines.add(ForgeI18n.parseMessage("fml.menu.mods.info.license", selectedMod.getOwningFile().getLicense()));
```
Comment (covers1624) [inline src/main/java/net/minecraftforge/client/gui/ModListScreen.java]: Why is this cast necessary? It would imply that there is a breaking change here requiring a cast.

File: fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractJarFileModProvider.java [hunk 9]
```diff
@@ -43,15 +32,4 @@ public void scanFile(final IModFile file, final Consumer<Path> pathConsumer) {
         }
         LOGGER.debug(LogMarkers.SCAN,"Scan finished: {}", file);
     }
-
-    @Override
-    public List<IModFile> scanMods() {
```
Comment (covers1624) [inline fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractJarFileModProvider.java]: The removal of this method causes a semantics change for anyone calling this function manually, as it will no longer scan mods, instead it will return an empty list, as it is not overridden anywhere else, the new overload is. Can we not fall back on this impl, or forward to `scanMods(Iterable)`?

Does this function actually get called anymore? What happens if someone has a custom mod locator? (ServerPackLocator is probably an example)

**Review by covers1624 — CHANGES_REQUESTED** [review]

**Review by covers1624 — APPROVED** [review]
